src/sig_stfl/lms/sig_stfl_lms_functions.c

```diff
@@ -0,0 +1,813 @@
+
+	struct hss_validate_inc ctx;
+	(void)hss_validate_signature_init(
+	    &ctx,               /* Incremental validate context */
+	    (const unsigned char *)pk,                /* Public key */
+	    (const unsigned char *)sm,
+	    (size_t)smlen,       /* Signature */
+	    0);                 /* Use the defaults for extra info */
+
+	(void)hss_validate_signature_update(
+	    &ctx,           /* Incremental validate context */
+	    (const void *) m,        /* Next piece of the message */
+	    (size_t)mlen);             /* Length of this piece */
+
+	bool status = hss_validate_signature_finalize(
+	                  &ctx,               /* Incremental validate context */
+	                  (const unsigned char *)sm,                /* Signature */
+	                  0);                 /* Use the defaults for extra info */
+
+	if (status) {
+		/* Signature verified */
+		return 0;
+	} else {
+		/* signature NOT verified */
+		return -1;
+	}
+}
+
+void oqs_secret_lms_key_free(OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL) {
+		return;
+	}
+
+	//TODO: cleanup lock_key
+
+	if (sk->secret_key_data) {
+		oqs_lms_key_data *key_data = (oqs_lms_key_data *)sk->secret_key_data;
+		if (key_data) {
+			OQS_MEM_secure_free(key_data->sec_key, key_data->len_sec_key);
+			key_data->sec_key = NULL;
+
+			OQS_MEM_secure_free(key_data->aux_data, key_data->len_aux_data);
+		}
+
+		OQS_MEM_insecure_free(key_data);
+		sk->secret_key_data = NULL;
+	}
+}
+
+/*
+ * Convert LMS secret key object to byte string
+ * Writes secret key + aux data if present
+ */
+OQS_STATUS oqs_serialize_lms_key(uint8_t **sk_key, size_t *sk_len, const OQS_SIG_STFL_SECRET_KEY *sk) {
+
+	if (sk == NULL || sk_len == NULL || sk_key == NULL) {
+		return OQS_ERROR;
+	}
+
+	oqs_lms_key_data *lms_key_data = sk->secret_key_data;
+
+	if (lms_key_data == NULL || lms_key_data->sec_key == NULL) {
+		return OQS_ERROR;
+	}
+
+	size_t key_len = lms_key_data->len_aux_data + lms_key_data->len_sec_key;
+
+	if (key_len == 0) {
+		return OQS_ERROR;
+	}
+
+	uint8_t *sk_key_buf = malloc(key_len * sizeof(uint8_t));
+
+	if (sk_key_buf == NULL) {
+		return OQS_ERROR;
+	}
+	/* pass back serialized data */
+	/*
+	 * Serialized data is sec_key followed by aux data
+	 * So aux data begins after buffer top + sec_key length
+	 */
+	if (lms_key_data->len_sec_key != 0) {
+		memcpy(sk_key_buf, lms_key_data->sec_key, lms_key_data->len_sec_key);
+	}
+
+	if (lms_key_data->len_aux_data != 0) {
+		memcpy(sk_key_buf + lms_key_data->len_sec_key, lms_key_data->aux_data, lms_key_data->len_aux_data);
+	}
+
+	*sk_key = sk_key_buf;
+	*sk_len = sk->length_secret_key + lms_key_data->len_aux_data;
+
+	return OQS_SUCCESS;
+}
+
+/*
+ * Convert LMS byte string to secret key object
+ * Writes secret key + aux data if present
+ * key_len is priv key length + aux length
+ */
+OQS_STATUS oqs_deserialize_lms_key(OQS_SIG_STFL_SECRET_KEY *sk, const size_t sk_len, const uint8_t *sk_buf, void *context) {
```

Comment:
it looks like `sk->sig` doesn't get initialized. Similar to the XMSS comment, this should probably be initialized with something since it's a required part of keygen (so it would be expected to be there).

Reply:
We never actually used `sk->sig` anywhere, so I just removed it.